streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -270,16 +272,20 @@ private void maybeAddNodeForOptimizationMetadata(final GraphNode node) {
 
     // use this method for testing only
     public void buildAndOptimizeTopology() {
-        buildAndOptimizeTopology(false);
+        buildAndOptimizeTopology(false, false);
     }
 
-    public void buildAndOptimizeTopology(final boolean optimizeTopology) {
+    public void buildAndOptimizeTopology(
+        final boolean optimizeTopology, final boolean optimizeSelfJoin) {
 
         mergeDuplicateSourceNodes();
         if (optimizeTopology) {
             LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
             optimizeKTableSourceTopics();
             maybeOptimizeRepartitionOperations();
+            if (optimizeSelfJoin) {
```

Comment:
This is a meta comment: we used a "two-pass" mechanism to apply optimization before since originally we did not pass in the `Configs` as part of the StreamsBuilder.build() process, and that two-pass mechanism has introduced a lot of issues and complicates (I think you've felt some at this time.. :P ) so the current plan is to get rid of the two-pass mechanism but try to optimize the topology in the first pass as we build the logical plan. So could you consider another approach than rewriting self joins if possible? In this way, we potentially would not need to consider awkward applicabilities like you have in https://github.com/apache/kafka/pull/12555/files#diff-ac1bf2b23b80784dec20b00fdc42f2df7e5a5133d6c68978fa44aea11e950c3aR398-R404 below.

Reply:
Hey @guozhangwang ! What do you mean by the  "two-pass" mechanism? Do you mean that one pass is building the plan and the second is optimizing it? What is wrong with doing this? 

The complications that I came across are actually not from this but from the fact that the logical plan is not so logical but has a lot of physical plan information already baked into it. I don't think it is a good idea to optimize the plan while building it (in one pass) as some optimizations cannot be identified until after the full plan is built. Moreover, optimizing the plan while building allows an optimization rule to only consider "local" information (up to the point of what the plan currently contains) and not the whole plan holistically which contributes to excluding optimizations that could otherwise be applied simply because the optimizer doesn't have all the information yet. Moreover, it contributes to "race" conditions between optimizations as applying one optimization might exclude another when the second could have been more beneficial. 

Regarding the optimization in this PR, it is not possible to identify whether it is a self-join until we have seen the entire plan. We need to make sure that it is the same source in both arguments of the join and that no transformation is applied to one argument and not the other. I am afraid, the checks that I am doing on whether a self-join is applicable are unavoidable whether we do the optimizations on the fly or afterward.

Reply:
Yeah I totally agree with you that today's logical plan is not so "logical" but have a lot of physical plan info baked..

Reply:
The main goal of the conditions is to ensure that the join arguments are the same and are not altered somewhere along the path from the root to the join in a unilateral (only one side) way.

The first condition is needed to make sure that the join has a single source ancestor. We don't want to apply the optimization if there is another source somewhere along the path from the root to the join node. This path might be long and might go through multiple joins etc. so just looking for a single parent is not enough. 

The second condition is needed to ensure that no transformation is applied to one side of the join and not the other.

The third condition is needed for the same reason because in my tests, I have seen topologies where the `ProcessorNodes` are added as siblings to the join and not as parents of it. For instance, in this example:
```
        final KStream<String, String> stream1 = builder.stream(Collections.singleton("t1"), consumed);
        stream1.mapValues(v -> v);
        final KStream<String, String> stream2 = builder.stream(Collections.singleton("t1"), consumed);
        stream1.join(stream2, MockValueJoiner.TOSTRING_JOINER, 
                             JoinWindows.ofTimeDifferenceWithNoGrace(ofMillis(100)));
```
the node for the `map` is a sibling of the join node and not a parent.

Reply:
Thanks @vpapavas , I think I start to understand some of your motivations here, but just to make sure I do, could we go over a list of examples below:

1.
```
stream1 = builder.stream("topic1");
stream1.join(stream1)
```
This case is optimizable, and just condition #2 alone should be sufficient to validate optimization to be applied.

2.
```
stream1 = builder.stream("topic1");
stream2 = builder.stream("topic1"); // same topic
stream1.join(stream2)
```
This case is optimizable, but since from the logical plan we still have two graph nodes we cannot validate the optimization.

3.
```
stream1 = builder.stream("topic1");
streams1.mapValues(v -> v);
stream2 = builder.stream("topic1"); // same topic
stream1.join(stream2)
```
This case is optimizable, but it seems our validation would exclude it?

4.
```
stream1 = builder.stream("topic1");
stream2 = builder.stream("topic2");
stream3 = stream1.merge(stream2);
stream3.join(stream3);
```
This case should be optimizable, but it seems our validation would exclude it?

Reply:
Hey @guozhangwang ! 
Cases 1 and 2 are optimizable and will be optimized by the algorithm. I have tests for this cases in `InternalStreamsBuilderTest`.  
Case 3 is not optimizable and won't be recognized. The reason is that processors like `mapValues` or `filter` or `transform` are black-boxes. There is no way to know how they change the contents of a stream hence there is no way to figure out if the two sides of the join are still the same. 
Case 4 could be optimizable but I did not consider it. I initially only had in-scope cases like 1 and 2. I can add it by adding a special rule to the rewriter that would check if the parent of the join is a merge, then it's ok to have multiple source nodes as long as they are ancestors of the merge node.

Reply:
Hmm.. for case 2), the logical plan would be like:

```
root -> source1
root -> source2
source1 -> join-window1
source2 -> join-window2
source1,source2 -> stream-join
```

Is that right?

For case 3), since we are not re-assigning the result of `mapValues` to stream` it should just be a no-op right? The logical plan would be like:

```
root -> source1
root -> source2
source1 -> join-window1
source1 -> map-values1
source2 -> join-window2
source1,source2 -> stream-join
```

Should that be considered optimizable as well?

Reply:
The logical plan will have only one source since they will get merged as they are referring to the same topic. Hence, all the nodes will be children of the source node and siblings of the join node. 
For example,
Case 2:
```
root -> source
source -> join-window1, join-window2, stream-join
```

and
Case 3:
```
root -> source
source -> join-window1, map-values1, join-window2, stream-join
```

Reply:
Discussed offline with @guozhangwang .

If the `map` or other transformation is a sibling of the join node but the join node has a single parent, then this means this transformation is a no-op and hence the optimization can be applied. For the transformation to not be a no-op, the join node must have two or more parents, hence the optimization won't be applicable.